File: app/src/main/java/activity/amigosecreto/repository/ParticipanteRepository.java
package activity.amigosecreto.repository;

import android.content.Context;

import androidx.annotation.VisibleForTesting;

import java.util.List;
import java.util.Map;

import activity.amigosecreto.db.Participante;
import activity.amigosecreto.db.ParticipanteDAO;

/**
 * Repository que encapsula todo o acesso a ParticipanteDAO.
 *
 * Isola Activities e ViewModels do SQLite, facilitando testes (mock/substituto)
 * e evolução futura (ex: Room, sync com servidor).
 *
 * Todos os métodos são síncronos e devem ser chamados a partir de uma thread de background.
 */
public class ParticipanteRepository {

    private final ParticipanteDAO dao;

    public ParticipanteRepository(Context context) {
        this.dao = new ParticipanteDAO(context);
    }

    @VisibleForTesting
    ParticipanteRepository(ParticipanteDAO dao) {
        this.dao = dao;
    }

    public List<Participante> listarPorGrupo(int grupoId) {
        dao.open();
        try {
            return dao.listarPorGrupo(grupoId);
        } finally {
            dao.close();
        }
    }

    public void inserir(Participante participante, int grupoId) {
        dao.open();
        try {
            dao.inserir(participante, grupoId);
        } finally {
            dao.close();
        }
    }

    public boolean atualizar(Participante participante) {
        dao.open();
        try {
            return dao.atualizar(participante);
        } finally {
            dao.close();
        }
    }

    public void remover(int id) {
        dao.open();
        try {
            dao.remover(id);
        } finally {
            dao.close();
        }
    }

    public void deletarTodosDoGrupo(int grupoId) {
        dao.open();
        try {
            dao.deletarTodosDoGrupo(grupoId);
        } finally {
            dao.close();
        }
    }

    public void limparSorteioDoGrupo(int grupoId) {
        dao.open();
        try {
            dao.limparSorteioDoGrupo(grupoId);
        } finally {
            dao.close();
        }
    }

    public void adicionarExclusao(int idParticipante, int idExcluido) {
        dao.open();
        try {
            dao.adicionarExclusao(idParticipante, idExcluido);
        } finally {
            dao.close();
        }
    }

    public void removerExclusao(int idParticipante, int idExcluido) {
        dao.open();
        try {
            dao.removerExclusao(idParticipante, idExcluido);
        } finally {
            dao.close();
        }
    }

    /**
     * Aplica todas as alterações de exclusão em uma única transação atômica (evita falha parcial
     * e múltiplos open/close que ocorreriam num loop de chamadas individuais).
     */
    public void salvarExclusoes(int participanteId, List<Integer> adicionar, List<Integer> remover) {
        dao.open();
        try {
            dao.salvarExclusoes(participanteId, adicionar, remover);
        } finally {
            dao.close();
        }
    }

    public boolean salvarSorteio(List<Participante> participantes, List<Participante> sorteados) {
        dao.open();
        try {
            return dao.salvarSorteio(participantes, sorteados);
        } finally {
            dao.close();
        }
    }

    public void marcarComoEnviado(int id) {
        dao.open();
        try {
            dao.marcarComoEnviado(id);
        } finally {
            dao.close();
        }
    }

    public String getNomeAmigoSorteado(int amigoId) {
        dao.open();
        try {
            return dao.getNomeAmigoSorteado(amigoId);
        } finally {
            dao.close();
        }
    }

    public Map<Integer, Integer> contarPorGrupo() {
        dao.open();
        try {
            return dao.contarPorGrupo();
        } finally {
            dao.close();
        }
    }

    public Participante buscarPorId(int id) {
        dao.open();
        try {
            return dao.buscarPorId(id);
        } finally {
            dao.close();
        }
    }
}

File: app/src/main/java/activity/amigosecreto/repository/ParticipanteRepository.kt
package activity.amigosecreto.repository

import androidx.annotation.VisibleForTesting
import activity.amigosecreto.db.Participante
import activity.amigosecreto.db.room.ParticipanteRoomDao

/**
 * Repository que encapsula todo o acesso a [ParticipanteRoomDao].
 *
 * Isola Activities e ViewModels do Room, facilitando testes via injeção do DAO.
 * Todos os métodos são suspend — devem ser chamados a partir de uma coroutine.
 *
 * A classe é `open` para que testes Kotlin possam criar subclasses anônimas com override.
 */
open class ParticipanteRepository @VisibleForTesting internal constructor(
    private val dao: ParticipanteRoomDao,
) {

    open suspend fun listarPorGrupo(grupoId: Int): List<Participante> =
        dao.listarPorGrupo(grupoId)

    open suspend fun inserir(participante: Participante, grupoId: Int) {
        participante.grupoId = grupoId
        val id = dao.inserir(participante)
        participante.id = id.toInt()
    }

    open suspend fun atualizar(participante: Participante): Boolean =
        dao.atualizar(participante) > 0

    open suspend fun remover(id: Int) = dao.remover(id)

    open suspend fun deletarTodosDoGrupo(grupoId: Int) = dao.deletarTodosDoGrupo(grupoId)

    open suspend fun limparSorteioDoGrupo(grupoId: Int) = dao.limparSorteioDoGrupo(grupoId)

    @VisibleForTesting(otherwise = VisibleForTesting.PACKAGE_PRIVATE)
    open suspend fun adicionarExclusao(idParticipante: Int, idExcluido: Int) {
        dao.salvarExclusoes(idParticipante, listOf(idExcluido), emptyList())
    }

    @VisibleForTesting(otherwise = VisibleForTesting.PACKAGE_PRIVATE)
    open suspend fun removerExclusao(idParticipante: Int, idExcluido: Int) {
        dao.salvarExclusoes(idParticipante, emptyList(), listOf(idExcluido))
    }

    open suspend fun salvarExclusoes(participanteId: Int, adicionar: List<Int>, remover: List<Int>) =
        dao.salvarExclusoes(participanteId, adicionar, remover)

    open suspend fun salvarSorteio(participantes: List<Participante>, sorteados: List<Participante>): Boolean =
        dao.salvarSorteio(participantes, sorteados)

    open suspend fun marcarComoEnviado(id: Int) = dao.marcarComoEnviado(id)

    open suspend fun confirmarCompra(id: Int) = dao.marcarConfirmacaoCompra(id)

    open suspend fun getNomeAmigoSorteado(amigoId: Int): String =
        dao.getNome(amigoId) ?: NOME_AMIGO_DESCONHECIDO

    open suspend fun contarPorGrupo(): Map<Int, Int> =
        dao.contarPorTodosGrupos().associate { it.grupoId to it.count }

    open suspend fun contarEnviadosPorGrupo(): Map<Int, Int> =
        dao.contarEnviadosPorTodosGrupos().associate { it.grupoId to it.count }

    open suspend fun buscarPorId(id: Int): Participante? = dao.buscarPorId(id)

    companion object {
        const val NOME_AMIGO_DESCONHECIDO = "Ninguém"
    }
}
